File: lex-audio/modules/lex.py
#!/usr/bin/env python3

from contextlib import closing
import configparser
import subprocess
import pyaudio
import boto3
import os

def make_recording():
    """Call bash script to record to 'request.wav'."""

    print("Listening...")
    subprocess.call("./record.sh")
    print("Done.")

def play_recording():
    """Call bash script to play 'request.wav'."""

    print("Playing 'recording.wav'...")
    subprocess.call("./play.sh")
    print("Done.")

def connect_to_lex():
    """Connect to Lex bot using AWS SDK."""

    config = configparser.ConfigParser()
    config.read('../../config.ini')

    if 'AWS' in config:
        ACCESS_KEY = config['AWS']['access_key']
        SECRET_KEY = config['AWS']['secret_key']

    print("Connecting to Reginald...")
    reginald = boto3.client(
        'lex-runtime',
        region_name='us-east-1',
        aws_access_key_id=ACCESS_KEY,
        aws_secret_access_key=SECRET_KEY
        )
    print("Done.")

    return reginald

def send_audio(reginald):
    """Send audio to Lex bot."""

    make_recording()
    play_recording()
    
    print("Posting content to Reginald...")
    response = reginald.post_content(
        botAlias='demo',
        botName='Reginald',
        userId='VoiceRecognition',
        contentType='audio/l16; rate=16000; channels=1',
        inputStream=open('./request.wav', 'rb')
        )
    print("Done.")

    print("Playing response...")
    with closing(response["audioStream"]) as stream:
        with open("response.mp3", "wb") as file:
                    file.write(stream.read())
    os.system("mpg123 response.mp3")
    print("Done.")

    print(response)

def send_text(reginald):
    """Send text to Lex bot."""

    print("Posting text to Reginald...")
    #Change input here.
    input = "tell me a joke"
    response = reginald.post_text(
        botName='Reginald',
        botAlias='demo',
        userId='TextRecognition',
        inputText=input
        )
    print("Done.")

    print("Printing response...")
    print(response['message'])
    print("Done.")

    print(response)

def main():
    """Main function for module."""

    #Define reginald object.
    reginald = connect_to_lex()

    #Uncomment ONE:
    send_audio(reginald)
    #send_text(reginald)

main()

File: lex-audio/modules/lex.js
// lex.js 
var AWS = require('aws-sdk'),
   fs = require('fs'),
   ts = require('tailstream'),
   exec = require('child_process').exec;

   var FULFILLED = 'Fulfilled',
   RESPONSE_FILE = 'response.mpeg',
   REMOVE_REQUEST_FILE = 'rm request.wav',
   RECORD_COMMAND = 'arecord --device=hw:1,0 --format S16_LE --rate 44100 -d 3 -c1 recording.wav',
   streaming = false,
   inputStream,
   lexruntime = new AWS.LexRuntime({
     region: 'us-east-1',
     credentials: new AWS.Credentials(
       'AKIAJJOUN2ESGFUQNI5A',
       'fGeo9g/qD2BGrnjXyhzuoS60DK6GtAnnEcWD7aHo', null)
   });   

var setupStream = function() {
   streaming = true;
   inputStream = ts.createReadStream('./request.wav');
   var params = {
     botAlias: '$LATEST',
     botName: 'Reginald',
     userId: 'lexHeadTesting',
     contentType: 'audio/l16; rate=16000; channels=1',
     inputStream: inputStream
   };

   lexruntime.postContent(params, function(err, data) {
     if (err) {
       console.log(err, err.stack);
       process.exit(1);
     } else {
       fs.writeFile(RESPONSE_FILE, data.audioStream, function(err) {
         if (err) {
           return console.log(err);
           process.exit(1);
         }
       });
       var playback = exec('sudo mpg321 ' + RESPONSE_FILE);
       playback.on('close', function(code) {
         exec('rm ' + RESPONSE_FILE);
         if (data.dialogState !== FULFILLED) {
           streaming = false;
           record();
         }
       });
     }
   });
 }

var record = function() {
   console.log('Listening...')
   var recording = exec(RECORD_COMMAND);
   recording.stderr.on('data', function(data) {
     console.log(data);
     if (!streaming) {
       setupStream();
     }
   });
   recording.on('close', function(code) {
     inputStream.done();
     exec(REMOVE_REQUEST_FILE);
   });
 }
record();